tables/endpointsecurity/src/fileeventstableplugin.cpp

```diff
@@ -0,0 +1,159 @@
+  IZeekLogger &logger;
+
+  RowList row_list;
+  std::mutex row_list_mutex;
+  std::size_t max_queued_row_count{0U};
+};
+
+Status FileEventsTablePlugin::create(Ref &obj,
+                                     IZeekConfiguration &configuration,
+                                     IZeekLogger &logger) {
+
+  try {
+    obj.reset(new FileEventsTablePlugin(configuration, logger));
+
+    return Status::success();
+
+  } catch (const std::bad_alloc &) {
+    return Status::failure("Memory allocation failure");
+
+  } catch (const Status &status) {
+    return status;
+  }
+}
+
+FileEventsTablePlugin::~FileEventsTablePlugin() {}
+
+const std::string &FileEventsTablePlugin::name() const {
+  static const std::string kTableName{"file_events"};
+
+  return kTableName;
+}
+
+const FileEventsTablePlugin::Schema &FileEventsTablePlugin::schema() const {
+
+  static const Schema kTableSchema = {
+      {"timestamp", IVirtualTable::ColumnType::Integer},
+      {"type", IVirtualTable::ColumnType::String},
+      {"parent_process_id", IVirtualTable::ColumnType::Integer},
+      {"orig_parent_process_id", IVirtualTable::ColumnType::Integer},
+      {"process_id", IVirtualTable::ColumnType::Integer},
+      {"user_id", IVirtualTable::ColumnType::Integer},
+      {"group_id", IVirtualTable::ColumnType::Integer},
+      {"platform_binary", IVirtualTable::ColumnType::Integer},
+      {"signing_id", IVirtualTable::ColumnType::String},
+      {"team_id", IVirtualTable::ColumnType::String},
+      {"cdhash", IVirtualTable::ColumnType::String},
+      {"path", IVirtualTable::ColumnType::String},
+      {"file_path", IVirtualTable::ColumnType::String}};
+
+  return kTableSchema;
+}
+
+Status FileEventsTablePlugin::generateRowList(RowList &row_list) {
+  std::lock_guard<std::mutex> lock(d->row_list_mutex);
+
+  row_list = std::move(d->row_list);
+  d->row_list = {};
+
+  return Status::success();
+}
+
+Status FileEventsTablePlugin::processEvents(
+    const IEndpointSecurityConsumer::EventList &event_list) {
+
+  for (const auto &event : event_list) {
+    Row row;
+
+    auto status = generateRow(row, event);
+    if (!status.succeeded()) {
+      return status;
+    }
+
+    if (!row.empty()) {
+      {
+        std::lock_guard<std::mutex> lock(d->row_list_mutex);
+        d->row_list.push_back(row);
+      }
+    }
+  }
+
+  if (d->row_list.size() > d->max_queued_row_count) {
+
+    auto rows_to_remove = d->row_list.size() - d->max_queued_row_count;
+
+    d->logger.logMessage(IZeekLogger::Severity::Warning,
+                         "file_events: Dropping " +
+                             std::to_string(rows_to_remove) +
+                             " rows (max row count is set to " +
+                             std::to_string(d->max_queued_row_count) + ")");
+
+    {
+      std::lock_guard<std::mutex> lock(d->row_list_mutex);
+      d->row_list.erase(d->row_list.begin(),
+                        std::next(d->row_list.begin(), rows_to_remove));
+    }
```

Comment:
Having the `if` outside the lock is fine if we compute `rows_to_remove` under the lock (`row_list`'s size might have changed, but getting the number of `row_to_remove` under the lock takes care of any races); I believe computing `row_to_remove` outside of the lock is not what we want, though (somebody could e.g., append more data in the meantime so we remove too little, or worse, data got evicted after computing it so we could end up removing too much data).

Let's move the computation under the lock and also add a safety check against overflow.
```suggestion

    std::lock_guard<std::mutex> lock(d->row_list_mutex);
    
    if (d->row_list.size() > d->max_queued_row_count) {
      auto rows_to_remove = d->row_list.size() - d->max_queued_row_count;

      d->logger.logMessage(IZeekLogger::Severity::Warning,
                           "file_events: Dropping " +
                               std::to_string(rows_to_remove) +
                               " rows (max row count is set to " +
                               std::to_string(d->max_queued_row_count) + ")");

      d->row_list.erase(d->row_list.begin(),
                        std::next(d->row_list.begin(), rows_to_remove));
    }

```